### PR TITLE
Module 4 code 

### PR DESCRIPTION
**Pull request by Yb1ku**
El material desarrollado en esta PR es el que se encuentra contenido en la carpeta `src/zrive/module4`. El resto de la organización será mejorado en futuras PR correspondientes a los módulos que haya que reorganizar. Los archivos principales incluidos en esta PR son: 

- `nonlinear_models_exploration.md`: Markdown con la exploración de modelos realizada y selección del que mejor rendimiento ha dado. 
- `data_processing.py`: Contiene funciones de utilidad para el tratamiento de los datos. 
- `train.py`: Script para entrenar el modelo escogido. 
- `inference.py`: Script para realizar inferencias al modelo entrenado. 
image
El mejor modelo encontrado es un XGBoost con los siguientes hiperparámetros: 

- **n_estimators**: 300 
- **eta**: 0.01 
- **gamma**: 5 
- **max_depth**: 8 

De momento no se ha realizado ningún análisis de la importancia de las variables. A partir de la clase de este miércoles, en la que entraremos en detalle en frameworks como SHAP, se volverá a realizar una PR sobre esta rama aplicando las técnicas de interpretabilidad aprendidas al modelo desarrollado aquí.

### CODE REVIEW

File: src/zrive/module3/data.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+from typing import Union, Tuple
+import pandas as pd
+
+
+def preprocess_data(data: pd.DataFrame,
+                    min_items: int = 5,
+                    train: bool = True) -> pd.DataFrame:
+    '''Preprocess the input data before splitting.'''
```
Comment (gscharly) [inline src/zrive/module3/data.py]: Si añades docstrings a las funciones, te sugiero que sigas algún estándar como [reStructuredText](https://peps.python.org/pep-0287/). Esto ayuda a describir la función, sus argumentos y su salida de una forma estandarizada.

Los docstrings ayudan cuando tu función va a ser utilizada en modo paquete por terceros, o cuando es una función muy reutilizada internamente en tu código

File: src/zrive/module4/inference.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        logger.info("Predictions made successfully.")
+        json_predictions = {}
+        for user in range(len(predictions)):
+            json_predictions[user_ids[user]] = predictions[user]
+
+        json_response = {
+            'statusCode': 200,
+            'body': json_predictions
+        }
+    except Exception as e:
```
Comment (gscharly) [inline src/zrive/module4/inference.py]: cuando se lanzan excepciones, es recomendable ser más específico con el tipo de excepción que capturas. Usar Exception puede complicar el análisis de errores, e incluso enmascararlos. Hay algún tipo de error que tu código lance? Quieres tratar diferentes errores de distinta forma?

**Review by gscharly — COMMENTED** [review]
Hola Diego,

- De nuevo, atento a la organización de los módulos. En esta PR se están metiendo archivos del módulo 3 (no se si se han colado o es a propósito). En cualquier caso, es buena práctica incluir en una PR solo los archivos relevantes a la misma.
- Las pruebas realizadas tienen sentido, empezando con un RF y terminando con un XGB. Sin embargo, echo de menos una optimización de hiperparámetros un poco más profunda, sobretodo en el caso de los dos primeros modelos. 
- Para los dos primeros algoritmos, me vuelve a faltar esa comparación con el conjunto de entrenamiento para comprobar el sobreajuste, cosa que si que realizas en el último algoritmo.
- Las implementaciones más "top" de Gradient Boosting Trees como XGBoost o Lightgbm tienen muchísimos hiperparámetros, y la optimización de los mismos no es trivial. He visto que usas un GridSearchCV que, en el caso de XGBoost, no es óptimo, ya que dejas mucho espacio de hiperparámetros sin explorar. Para el futuro, cuando trabajes con estos algoritmos, técnicas de optimización basadas en estadística bayesiana funcionan bastante bien. Le puedes echar un ojo a esta [librería](https://github.com/bayesian-optimization/BayesianOptimization)
- Me falta alguna hipótesis de por qué ciertas pruebas no funcionan. En la primera prueba con GBT, por qué crees que aumentar el número de estimadores no mejora las métricas, como si lo hace en RF? Qué efecto percibes al aumentar tanto la tasa de aprendizaje y qué crees que lo causa?
- En tu notebook podrías intentar importar las funciones genéricas que has implementado en tus módulos Python, para evitar tener código duplicado.
- Espero ese análisis de variable en una futura PR!

**Comment by Yb1ku** [comment]
Hola, @gscharly: 

He modificado el archivo `train.py` para que incluya más parámetros de entrada en la función `fit_model`. He cambiado también las docstrings en todos los archivos .py para mayor claridad. 

Por otro lado, el archivo `model_analysis.md` contiene un análisis SHAP del modelo XGBoost entrenado al inicio de la PR. Después de ver qué variables son las más importantes según SHAP, he ido entrenando versiones reducidas del modelo (con menos variables), para tratar de encontrar una versión más ligera con el mismo rendimiento. Al final, usando las 13 variables más importantes (en vez de las 22 que usa el modelo original), se obtiene el mismo rendimiento. Sin embargo, me parece que el modelo optimizado es menos interpretable. Este es el beeswarm del modelo original: 
image
Y este el del modelo reducido: 
image
En el modelo original se ve de forma más clara la influencia de las variables. Sin embargo, el reducido, pese a que `ordered_before` mantiene el mismo comportamiento (aunque la magnitud de su influencia es distinta), `abandoned_before`, `set_as_regular` y `active_snoozed` ya no hacen lo mismo. Tengo dos hipótesis acerca de esto: 

1. El modelo ha tenido que reorganizar cómo usa las variables. Esta disminución en la interpretabilidad de algunas variables es algún tipo de "precio a pagar" por tener mayor eficiencia. 
2. He hecho algo mal/diferente con los datos empleados para sacar los beeswarms. 

Por último, con respecto al tema de las exceptions, he quitado lo que tenía puesto. La duda que tengo con ello es que entiendo que, si por lo que sea hay un error, sería bueno que el código devolviese algo, aunque sea un status code 4xx/5xx y el mensaje de error. ¿Se debe ir especificando para cada tipo de error qué quieres que devuelva, o solo poner unos cuantos tipos de errores? Y siguiendo con eso, ¿cómo se debería enfocar los tests unitarios de esa parte, asegurándote que devuelve los status codes correctos? 

Muchas gracias y un saludo.

**Comment by gscharly** [comment]
Hola Diego,

Perdona por la tardanza.

Buen análisis, es positivo intentar conseguir el modelo más simple sin empeorar las métricas de clasificación. En cuanto al cambio de importancia de variables, puede tener varias razones. ¿Has reentrenado con los mismos hiperparámetros, o al elegir menos variables has vuelto a optmizar hiperparámetros? Al ser un modelo basado en árboles de decisión, las variables que se elijen pueden cambiar. Además, si estás variables están relacionadas entre sí, es posible que su importancia se "esconda" detrás de otra variable. También hay que tener en cuenta el tipo de variables (son todas continuas, hay alguna binaria?)

Respecto a las excepciones, es buena práctica ser lo más específico posible con ellas. Si sabes que tu código puede lanzar un error determinado, captura esa excepción específica con su tipo, no con un Exception genérico. Los códigos de error (4xx, 5xx) son necesarios cuanto tu código va a ser usado por terceros, sobretodo cuando se expone a través de APIs. Internamente puedes manejar los errores con excepciones de Python. Respecto a los tests, si tu código trata excepciones, siempre es útil crear casos de test en los que se fuercen esas excepciones. Herrramientas como pytest proporcionan formas sencillas de trabajar con excepciones.
